### PR TITLE
docs: klip-54: Optimized Range Pull Queries on Tables

### PR DESCRIPTION
**Pull request by patrickstuedi**
### Description 
_What behavior do you want to change, why, how does your patch achieve the changes?_

### Testing done 
_Describe the testing strategy. Unit and integration tests are expected for any behavior changes._

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: design-proposals/klip-54-optimized-range-pull-queries.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+**Status**: _In Discussion_ | 
+**Discussion**: https://github.com/confluentinc/ksql/pull/7993
+
+**tl;dr:** For range pull queries on tables, use the range interface provided by the state store for retrieving records instead of doing a table scan.
+           
+## Motivation and background
+
+Currently, range queries on primary keys (e.g., WHERE colX < T) are implemented using a table scan followed by a filtering operation, which is inefficient. Recently, range scan capabilities were added to Kafka streams (#4064). Following that work, we should optimize the execution of range queries in ksqlDB by directly leveraging the dedicated range query interface to improve I/O efficiency and performance. 
+
+## What is in scope
```
Comment (agavra) [inline design-proposals/klip-54-optimized-range-pull-queries.md:14]: I think this is a good place to discuss the implications of serdes on range scans, and how we eventually plan to support non-string range scans

Comment (patrickstuedi) [reply design-proposals/klip-54-optimized-range-pull-queries.md:14]: Added a paragraph in the "What is not in this scope" about the String restriction and how we can support binary types in the future.

**Review by agavra — APPROVED** [review]
LGTM! thanks @patrickstuedi

File: design-proposals/klip-54-optimized-range-pull-queries.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+## Public APIS
+
+No new syntax is added. 
+
+## Design
+
+In order to efficiently support range queries on tables, several modifications have to be made with regard to how opportunities for range scans are detected, how the physical plan is generated and how materialization is implemented:
+1. Generate KeyConstrains for range expressions (instead of NoKeyConstraints)
+2. Introduce a new physical operator called RangeScanOperator to handle queries consisting of KeyConstraints of type "<", "<=", ">", ">="
+3. Add support for range scans in ksqlDB materialization
```
Comment (vvcephei) [inline design-proposals/klip-54-optimized-range-pull-queries.md:37]: I think your current plan is a little different now, right?

**Review by vvcephei — APPROVED** [review]
Thanks, @patrickstuedi !

I made some suggestions, but I'm fundamentally in favor.

File: design-proposals/klip-54-optimized-range-pull-queries.md (hunk 2 above)
Comment (patrickstuedi) [reply design-proposals/klip-54-optimized-range-pull-queries.md]: Right, thanks for spotting. Changed the text..

File: design-proposals/klip-54-optimized-range-pull-queries.md [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+Currently, range queries on primary keys (e.g., WHERE colX < T) are implemented using a table scan followed by a filtering operation, which is inefficient. Recently, range scan capabilities were added to Kafka streams (#4064). Following that work, we should optimize the execution of range queries in ksqlDB by directly leveraging the dedicated range query interface to improve I/O efficiency and performance. 
+
+## What is in scope
+
+* Pull queries on tables with a WHERE clause including ">", ">=", "<", "<=" expressions should be executed using the range interface in the state store. 
+
+## What is not in scope
+
+* There exists a potential for optimizing pull queries with conjunctions or disjunctions of range expressions, e.g., by compacting multiple range expressions into fewer expressions, leading to fewer range calls on the state store and, thus, better I/O efficiency and improved performance. Such optimizations are not part of this KLIP.
```
[the remark this answers is not in the document]
Comment (patrickstuedi) [reply design-proposals/klip-54-optimized-range-pull-queries.md]: Good point

**Pull request merged**